File: fsharp/robot-simulator/RobotSimulator.fs
module RobotSimulator

type Bearing = 
    | North
    | East
    | South
    | West

type Robot = {
    bearing: Bearing;
    position: int * int;
}
     
let createRobot bearing coordinate = { 
    bearing = bearing
    position = coordinate
}

let turnLeft robot =
    let nextBearing = 
        match robot.bearing with
        | Bearing.East -> Bearing.North
        | Bearing.West -> Bearing.South
        | Bearing.North -> Bearing.West
        | Bearing.South -> Bearing.East
    createRobot nextBearing robot.position

let turnRight robot = 
    let nextBearing = 
        match robot.bearing with
        | Bearing.East  -> Bearing.South
        | Bearing.West  -> Bearing.North
        | Bearing.North -> Bearing.East
        | Bearing.South -> Bearing.West
    createRobot nextBearing robot.position

let advance robot = 
    let (x, y) = robot.position
    let nextPosition = 
        match robot.bearing with
        | Bearing.East -> (x+1, y)
        | Bearing.West -> (x-1, y)
        | Bearing.North -> (x, y+1)
        | Bearing.South -> (x, y-1)
    createRobot robot.bearing nextPosition

let move robot command =
    match command with
    | 'A' -> advance robot
    | 'L' -> turnLeft robot
    | 'R' -> turnRight robot
    | _ -> failwith "Unknown command"

let simulate robot instructions =
    let mutable movedRobot = robot
    let iterator command = movedRobot <- move movedRobot command
    instructions |> String.iter iterator
    movedRobot

File: python/robot-simulator/robot_simulator.py
NORTH = 0
EAST = 1
SOUTH = 2
WEST = 3


class Robot(object):

    def __init__(self, bearing=NORTH, x=0, y=0):
        self.bearing = bearing
        self.x = x
        self.y = y

    @property
    def coordinates(self):
        return self.x, self.y

    def turn_right(self):
        self.bearing = (self.bearing + 1) % 4

    def turn_left(self):
        self.bearing = (self.bearing - 1) % 4

    def advance(self):
        if self.bearing == NORTH:
            self.y += 1
        elif self.bearing == EAST:
            self.x += 1
        elif self.bearing == SOUTH:
            self.y -= 1
        elif self.bearing == WEST:
            self.x -= 1

    def simulate(self, steps):
        for step in steps:
            if step == 'A':
                self.advance()
            elif step == 'L':
                self.turn_left()
            elif step == 'R':
                self.turn_right()

